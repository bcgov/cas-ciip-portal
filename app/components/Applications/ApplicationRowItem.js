import React, {Component} from 'react';
import propTypes from 'prop-types';
import {Button, Badge} from 'react-bootstrap';

class ApplicationRowItem extends Component {
  static propTypes = {
    application: propTypes.shape({
      applicationId: propTypes.number,
      applicationStatus: propTypes.string,
      submissionDate: propTypes.string,
      facilityName: propTypes.string,
      operatorName: propTypes.string,
      reportingYear: propTypes.string,
      bcghgid: propTypes.string
    }).isRequired
  };

  render() {
    const {application} = this.props;
    console.log('my application', application);
    const statusBadgeColor = {
      attention: 'warning',
      pending: 'info',
      declined: 'danger',
      approved: 'success'
    };
    // Const url = `http://localhost:3000/public/dashboard/985719f1-7eae-4c49-88a9-7d6c8edc1ad4?application_id=${application.applicationId}`;
    // Const url = `https://metabase-wksv3k-dev.pathfinder.gov.bc.ca/public/dashboard/bb6a4b75-3a7f-4fab-9268-cb013ecfcb7b?application_id=${application.applicationId}`;
    const applicationDetails = `/application-details?application_id=${application.applicationId}&reportingyear=${application.reportingYear}&bcghgid=${application.bcghgid}`;

    return (
      <tr>
        <td>{application.applicationId}</td>
        <td>{application.operatorName}</td>
        <td>{application.facilityName}</td>
        <td>{application.submissionDate}</td>
        <td>
          <Badge
            pill
            style={{width: '100%'}}
            variant={statusBadgeColor[application.applicationStatus]}
          >
            {application.applicationStatus}
          </Badge>
        </td>
        <td>
          <Button href={applicationDetails} target="_blank" variant="primary">
            View Application
          </Button>
        </td>
      </tr>
    );
  }
}

export default ApplicationRowItem;
