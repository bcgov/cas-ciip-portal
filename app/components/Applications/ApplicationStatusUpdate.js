import React, {Component} from 'react';
import propTypes from 'prop-types';


class ApplicationStatusUpdate extends Component {
  static propTypes = {
    applicationStatus: propTypes.string.isRequired,
    setApplicationStatus: propTypes.func.isRequired
  };

  render() {
    const {applicationStatus, setApplicationStatus} = this.props;
    const statusBadgeColor = {
      pending: 'info',
      attention: 'warning',
      declined: 'danger',
      approved: 'success'
    };

    
}

export default ApplicationStatusUpdate;
