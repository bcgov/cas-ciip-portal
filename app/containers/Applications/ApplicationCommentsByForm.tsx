import React from 'react';
import {createFragmentContainer, graphql, RelayProp} from 'react-relay';
import moment from 'moment';

/*
 * The ApplicationComments renders all the comments on the various sections of the application
 */

interface Props {
  applicationReview;
  relay: RelayProp;
}

export const ApplicationCommentsByForm: React.FunctionComponent<Props> = props => {
  const {applicationReview} = props;
  const reviewComments = applicationReview
    ? applicationReview.reviewCommentsByApplicationReviewId.edges
    : null;

  return (
    <>
      <div key={applicationReview.id} className="review-box">
        <div>
          Set to <strong>{applicationReview.reviewStatus}</strong>
          &nbsp;
          <small>
            {moment(applicationReview.createdAt).format(
              'MMM Do YYYY, h:mm:ss a'
            )}
          </small>
        </div>
        <ul className="comment-box">
          {reviewComments.map(({node}) => (
            <li key={node.id} className="comments">
              <div>{node.description}</div>
              <div>
                <small>
                  {moment(node.createdAt).format('MMM Do YYYY, h:mm:ss a')}
                </small>
              </div>
            </li>
          ))}
        </ul>
      </div>

      <style jsx>{`
        .form-result-box .review-box {
          border-bottom: 1px solid #d2d2d2;
          margin-bottom: 20px;
        }
      `}</style>
    </>
  );
};

export default createFragmentContainer(ApplicationCommentsByForm, {
  applicationReview: graphql`
    fragment ApplicationCommentsByForm_applicationReview on ApplicationReview {
      id
      reviewStatus
      formResultId
      createdAt
      reviewCommentsByApplicationReviewId(first: 2147483647)
        @connection(
          key: "ApplicationCommentsByForm_reviewCommentsByApplicationReviewId"
        ) {
        edges {
          node {
            id
            description
            createdAt
          }
        }
      }
    }
  `
});
