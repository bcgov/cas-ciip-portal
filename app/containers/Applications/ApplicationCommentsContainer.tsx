import React, {useState} from 'react';
import {Collapse, Table, Button} from 'react-bootstrap';
import {createFragmentContainer, graphql, RelayProp} from 'react-relay';
import {ApplicationCommentsContainer_formResult} from 'ApplicationCommentsContainer_formResult.graphql';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faChevronDown, faChevronRight} from '@fortawesome/free-solid-svg-icons';
import ApplicationCommentsBox from './ApplicationCommentsByForm';

/*
 * The ApplicationComments renders all the comments on the various sections of the application
 */

interface Props {
  formResult: ApplicationCommentsContainer_formResult;
  relay: RelayProp;
  review: boolean;
}

export const ApplicationCommentsComponent: React.FunctionComponent<Props> = (
  props
) => {
  const {formResult, review} = props;
  const [isOpen, setIsOpen] = useState(false);
  const [showResolved, toggleShowResolved] = useState(false);
  return (
    <>
      <div key={formResult.id} className="form-result-box">
        <h3>
          {formResult.formJsonByFormId.name}{' '}
          <i
            aria-label="Collapse or expand comment"
            title="Collapse or expand comment"
            style={{float: 'right', marginRight: '10px'}}
            onClick={() => setIsOpen(!isOpen)}
          >
            {isOpen ? (
              <FontAwesomeIcon icon={faChevronRight} />
            ) : (
              <FontAwesomeIcon icon={faChevronDown} />
            )}
          </i>
        </h3>
        <Collapse in={!isOpen}>
          <div>
            <Table hover>
              <thead className="comments-header">
                <tr>
                  <th>General Comment</th>
                  {review ? (
                    <th style={{textAlign: 'right'}}>Resolve/Delete</th>
                  ) : null}
                </tr>
              </thead>
              <tbody>
                {formResult.internalGeneralComments.edges.map(({node}) => {
                  if (node.resolved && !showResolved) return null;
                  if (
                    !review &&
                    formResult.applicationByApplicationId
                      .applicationRevisionStatus.applicationRevisionStatus ===
                      'SUBMITTED'
                  )
                    return null;
                  return (
                    <ApplicationCommentsBox
                      key={node.id}
                      review={review}
                      reviewComment={node}
                      version={formResult.versionNumber}
                      formResultId={formResult.id}
                    />
                  );
                })}
              </tbody>
            </Table>
            <Table hover>
              <thead className="comments-header">
                <tr>
                  <th>Request Changes</th>
                  {review ? (
                    <th style={{textAlign: 'right'}}>Resolve/Delete</th>
                  ) : null}
                </tr>
              </thead>
              <tbody>
                {formResult.requestedChangeComments.edges.map(({node}) => {
                  if (node.resolved && !showResolved) return null;
                  if (
                    !review &&
                    formResult.applicationByApplicationId
                      .applicationRevisionStatus.applicationRevisionStatus ===
                      'SUBMITTED'
                  )
                    return null;
                  return (
                    <ApplicationCommentsBox
                      key={node.id}
                      review={review}
                      reviewComment={node}
                      version={formResult.versionNumber}
                      formResultId={formResult.id}
                    />
                  );
                })}
              </tbody>
            </Table>
            <Button
              variant="link"
              style={{textDecoration: 'none'}}
              onClick={() => toggleShowResolved(!showResolved)}
            >
              {showResolved ? 'Hide Resolved' : 'Show Resolved'}
            </Button>
          </div>
        </Collapse>
      </div>
      <style jsx global>{`
        .form-result-box {
          padding: 25px;
          margin-bottom: 20px;
          border: 1px solid #ff713a;
          border-radius: 5px;
          background-color: #fff4dcab;
          font-size: 14px;
        }

        .form-result-box h3 {
          border-bottom: 1px solid #888;
          padding-bottom: 10px;
          margin-bottom: 20px;
          font-size: 1.25rem;
        }

        .hide-title label.form-label {
          display: none;
        }
        i {
          cursor: pointer;
        }
        .comments-header {
          font-weight: bold;
          color: #036;
        }
        .comments-header th {
          border: 0;
        }
      `}</style>
    </>
  );
};

export default createFragmentContainer(ApplicationCommentsComponent, {
  formResult: graphql`
    fragment ApplicationCommentsContainer_formResult on FormResult {
      id
      versionNumber
      applicationByApplicationId {
        id
        rowId
        applicationRevisionStatus(versionNumberInput: $version) {
          applicationRevisionStatus
        }
      }
      formJsonByFormId {
        rowId
        name
      }
      internalGeneralComments(first: 2147483647)
        @connection(
          key: "ApplicationCommentsContainer_internalGeneralComments"
        ) {
        edges {
          node {
            id
            resolved
            ...ApplicationCommentsByForm_reviewComment
          }
        }
      }
      requestedChangeComments(first: 2147483647)
        @connection(
          key: "ApplicationCommentsContainer_requestedChangeComments"
        ) {
        edges {
          node {
            id
            resolved
            ...ApplicationCommentsByForm_reviewComment
          }
        }
      }
    }
  `
});
