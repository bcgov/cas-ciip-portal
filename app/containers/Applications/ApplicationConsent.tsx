import React from 'react';
import {createFragmentContainer, graphql, RelayProp} from 'react-relay';
import {useRouter} from 'next/router';
import {Button} from 'react-bootstrap';
import updateApplicationRevisionMutation from 'mutations/application/updateApplicationRevisionMutation';
import {ApplicationConsent_application} from 'ApplicationConsent_application.graphql';

interface Props {
  application: ApplicationConsent_application;
  relay: RelayProp;
}

export const ApplicationConsent: React.FunctionComponent<Props> = (props) => {
  const router = useRouter();

  const handleContinueClick = async () => {
    const {environment} = props.relay;

    const variables = {
      input: {
        id: props.application.latestDraftRevision.id,
        applicationRevisionPatch: {
          legalDisclaimerAccepted: true,
          versionNumber: props.application.latestDraftRevision.versionNumber
        }
      }
    };

    await updateApplicationRevisionMutation(environment, variables);

    router.push({
      pathname: '/reporter/application',
      query: {
        applicationId: props.application.id,
        version: props.application.latestDraftRevision.versionNumber
      }
    });
  };

  return (
    <Button variant="primary" size="lg" onClick={handleContinueClick}>
      Consent and continue
    </Button>
  );
};

export default createFragmentContainer(ApplicationConsent, {
  application: graphql`
    fragment ApplicationConsent_application on Application {
      id
      latestDraftRevision {
        id
        versionNumber
      }
    }
  `
});
