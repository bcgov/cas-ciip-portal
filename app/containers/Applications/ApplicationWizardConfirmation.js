import React from 'react';

const ApplicationWizardConfirmation = () => {
  return (
    <>
      <p>
        TODO: This is where we display a summary of the complete application and
        ask you to submit it.
      </p>
    </>
  );
};

export default ApplicationWizardConfirmation;
