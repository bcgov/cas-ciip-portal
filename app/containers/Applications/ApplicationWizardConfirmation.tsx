import React, { useState, useMemo } from "react";
import { Card, Alert } from "react-bootstrap";
import validateJsonSchema from "@rjsf/core/dist/cjs/validate";
import { createFragmentContainer, graphql, RelayProp } from "react-relay";
import SubmitApplication from "components/SubmitApplication";
import { ApplicationWizardConfirmation_query } from "ApplicationWizardConfirmation_query.graphql";
import { ApplicationWizardConfirmation_applicationRevision } from "ApplicationWizardConfirmation_applicationRevision.graphql";
import ApplicationDetailsContainer from "./ApplicationDetailsContainer";
import ApplicationOverrideJustification from "components/Application/ApplicationOverrideJustification";
import type { FormJson } from "types";
import ScrollableApplicationDisclaimer from "components/Application/ScrollableApplicationDisclaimer";

/*
 * The ApplicationWizardConfirmation renders a summary of the data submitted in the application,
 * and allows the user to submit their application.
 */

interface Props {
  query: ApplicationWizardConfirmation_query;
  applicationRevision: ApplicationWizardConfirmation_applicationRevision;
  relay: RelayProp;
}

export const ApplicationWizardConfirmationComponent: React.FunctionComponent<Props> = ({
  applicationRevision,
  query,
}) => {
  const hasErrors = useMemo(() => {
    return applicationRevision.orderedFormResults.edges.some(
      ({ node: { formResult, formJsonByFormId } }) => {
        const { schema, customFormats } = formJsonByFormId.formJson as FormJson;

        const { errors } = validateJsonSchema(
          formResult,
          schema,
          undefined,
          undefined,
          undefined,
          customFormats
        );
        return errors.length > 0;
      }
    );
  }, [applicationRevision]);

  const [overrideActive, setOverrideActive] = useState(
    applicationRevision.overrideJustification !== null
  );
  const [applicationDetailsRendered, setApplicationDetailsRendered] = useState(
    false
  );

  const hasValidationWarnings = applicationRevision.validation.edges.some(
    ({ node }) => node.isOk === false
  );

  const validationAlert = (
    <Alert variant="warning">
      <p>
        <strong>
          Some inconsistencies were found in your application. Please address
          these items before submitting by correcting the reported values or
          adding a comment in the relevant section to provide an explanation.
        </strong>
      </p>
      <ul>
        {applicationRevision.validation.edges
          .filter(({ node }) => node.isOk === false)
          .map(({ node }) => (
            <li key={node.validationDescription}>
              {node.validationFailedMessage}
            </li>
          ))}
      </ul>
    </Alert>
  );

  // Ensure the application being submitted is either for the correct reporting year (i.e. not a previous year), or a version greater than 1 (i.e. an existing application being re-submitted)
  const showSubmitButton =
    applicationRevision.applicationByApplicationId.reportingYear ===
      query.openedReportingYear?.reportingYear ||
    applicationRevision.versionNumber > 1;

  return (
    <>
      <h1>Summary of your application:</h1>
      <p style={{ fontSize: "1.25rem", fontWeight: 500 }}>
        Please review the information you have provided before continuing.
      </p>
      {hasValidationWarnings && validationAlert}
      <br />
      <ApplicationOverrideJustification
        overrideActive={overrideActive}
        setOverrideActive={setOverrideActive}
        applicationOverrideJustification={
          applicationRevision.overrideJustification
        }
        revisionId={applicationRevision.id}
        hasErrors={hasErrors}
        applicationDetailsRendered={applicationDetailsRendered}
      />
      <ApplicationDetailsContainer
        liveValidate
        query={query}
        applicationRevision={applicationRevision}
        review={false}
        setApplicationDetailsRendered={setApplicationDetailsRendered}
        diffQuery={null}
      />
      <br />
      {hasErrors && !overrideActive ? (
        <div className="errors">
          Your Application contains errors that must be fixed before submission.
        </div>
      ) : (
        <>
          <Card id="next-step">
            <Card.Header>
              <h2>Before you submit</h2>
            </Card.Header>
            <Card.Body>
              <p>
                Thank you for reviewing your application to the CleanBC
                Industrial Incentive Program.
              </p>
              <p>
                Your application is almost complete. Prior to submitting your
                application, please review and accept the following terms below.
              </p>
              <p>
                By submitting the application you agree that the information
                contained in this application, or information contained in
                emissions reports submitted under the{" "}
                <a
                  href="https://www.bclaws.ca/civix/document/id/complete/statreg/14029_01"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <em>Greenhouse Gas Industrial Reporting and Control Act</em>
                </a>
                , may be disclosed to British Columbia government employees,
                contractors and agencies for the purpose of administering the
                CleanBC Program for Industry or the{" "}
                <a
                  href="https://www.bclaws.ca/civix/document/id/complete/statreg/14029_01"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <em>Greenhouse Gas Industrial Reporting and Control Act</em>
                </a>
                .
              </p>
              <ScrollableApplicationDisclaimer />
            </Card.Body>
          </Card>
          <br />
          {showSubmitButton && (
            <SubmitApplication applicationRevision={applicationRevision} />
          )}
        </>
      )}
      <style jsx global>
        {`
          h1 {
            font-size: 30px;
          }
          .errors {
            margin-left: 20px;
            padding: 20px;
            background: #ce5c5c;
            color: white;
            font-size: 20px;
          }
          @media print {
            .print-hide {
              display: none !important;
            }
          }
        `}
      </style>
    </>
  );
};

export default createFragmentContainer(ApplicationWizardConfirmationComponent, {
  applicationRevision: graphql`
    fragment ApplicationWizardConfirmation_applicationRevision on ApplicationRevision {
      id
      overrideJustification
      versionNumber
      ...SubmitApplication_applicationRevision
      ...ApplicationDetailsContainer_applicationRevision
      orderedFormResults {
        edges {
          node {
            formResult
            formJsonByFormId {
              formJson
            }
          }
        }
      }
      validation {
        edges {
          node {
            validationDescription
            validationFailedMessage
            isOk
          }
        }
      }
      applicationByApplicationId {
        reportingYear
      }
    }
  `,
  query: graphql`
    fragment ApplicationWizardConfirmation_query on Query {
      ...ApplicationDetailsContainer_query
      openedReportingYear {
        reportingYear
      }
    }
  `,
});
