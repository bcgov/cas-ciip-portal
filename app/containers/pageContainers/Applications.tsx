import React, {Component} from 'react';
import {graphql} from 'react-relay';
import {ApplicationsQueryResponse} from '__generated__/ApplicationsQuery.graphql';
import DefaultLayout from '../../layouts/default-layout';
import ApplicationListContainer from '../Applications/ApplicationListContainer';

interface Props {
  query?: ApplicationsQueryResponse['query'];
}

class Applications extends Component<Props> {
  // In downstream component
  // onClick={(event) => handleApplicationsAction({action: Applications.SORT_ACTION})})
  // I didn't fully understand how to make this work, so I moved on with the way I did.
  static query = graphql`
    query ApplicationsQuery(
      $orderByField: String
      $direction: String
      $searchField: String
      $searchValue: String
    ) {
      query {
        session {
          ...Header_session
        }
        ...ApplicationListContainer_query
          @arguments(
            orderByField: $orderByField
            direction: $direction
            searchField: $searchField
            searchValue: $searchValue
          )
      }
    }
  `;

  state = {
    orderByField: 'operator_name',
    direction: 'ASC',
    orderByDisplay: 'Operator Name',
    searchField: null,
    searchValue: null,
    searchDisplay: 'No Filter'
  };

  static async getInitialProps() {
    return {
      variables: {
        orderByField: 'operator_name',
        direction: 'ASC',
        searchField: null,
        searchValue: null
      }
    };
  }

  toggleDirection = () => {
    this.state.direction === 'ASC'
      ? this.setState({direction: 'DESC'})
      : this.setState({direction: 'ASC'});
  };

  sortApplications = (event, eventKey) => {
    this.setState({
      orderByField: eventKey,
      orderByDisplay: event.target.text
    });
  };

  applySearchField = (event, eventKey) => {
    this.setState({
      searchField: eventKey,
      searchDisplay: event.target.text,
      searchValue: null
    });
  };

  applySearchValue = event => {
    if (this.state.searchField !== 'none') {
      this.setState({searchValue: event.nativeEvent.target[0].value});
    }
  };

  handleEvent = (event, eventKey) => {
    event.preventDefault();
    event.stopPropagation();
    event.persist();
    this[event.target.id](event, eventKey);
  };

  render() {
    const {query} = this.props;
    return (
      <>
        <DefaultLayout title="Applications" session={query.session}>
          <ApplicationListContainer
            query={query}
            orderByDisplay={this.state.orderByDisplay}
            searchDisplay={this.state.searchDisplay}
            direction={this.state.direction}
            orderByField={this.state.orderByField}
            searchField={this.state.searchField}
            searchValue={this.state.searchValue}
            handleEvent={this.handleEvent}
          />
        </DefaultLayout>
      </>
    );
  }
}

export default Applications;
