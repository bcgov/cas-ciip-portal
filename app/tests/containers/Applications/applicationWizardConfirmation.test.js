import React from 'react';
import {shallow} from 'enzyme';
import {ApplicationWizardConfirmationComponent} from '../../../containers/Applications/ApplicationWizardConfirmation';

describe('ApplicationWizardConfirmationComponent', () => {
  it('should render the summary confirmation component', () => {
    const query = {
      application: {
        formResultsByApplicationId: {
          edges: [{node: {formResult: '{"id": "form-results"}'}}]
        },
        orderedFormResults: {
          edges: [{node: {id: 'abc'}}]
        }
      }
    };
    const r = shallow(<ApplicationWizardConfirmationComponent query={query} />);
    expect(r).toMatchSnapshot();
    expect(
      r
        .find('Relay(ApplicationWizardConfirmationCardItemComponent)')
        .prop('formTitle')
    ).toBe('id');
  });
});
