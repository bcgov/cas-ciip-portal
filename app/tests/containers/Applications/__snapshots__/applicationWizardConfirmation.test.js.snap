// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationWizardConfirmationComponent should render the summary confirmation component 1`] = `
<Fragment>
  <h1>
    Summary of your application:
  </h1>
  <h5>
    Please review the information you have provided before continuing.
  </h5>
  <br />
  <Relay(ApplicationWizardConfirmationCardItemComponent)
    formSubtitle="id"
    formTitle="id"
    key="id id"
    resultObject={
      Object {
        "id": Object {
          "id": "form-results",
        },
      }
    }
  />
  <Link
    href={
      Object {
        "pathname": "/ciip-application",
        "query": Object {
          "applicationId": undefined,
          "certificationPage": true,
          "formResultId": "abc",
        },
      }
    }
    passHref={true}
  >
    <Button
      active={false}
      className="float-right"
      disabled={false}
      style={
        Object {
          "marginTop": "10px",
        }
      }
      type="button"
      variant="primary"
    >
      Next
    </Button>
  </Link>
</Fragment>
`;
